packages/aws-cdk/lib/contextplugins.ts

```diff
@@ -48,21 +53,85 @@ export class SSMContextProviderPlugin implements ContextProviderPlugin {
   }
 }
 
+export class HostedZoneContextProviderPlugin implements ContextProviderPlugin {
+
+  constructor(private readonly aws: SDK) {
+  }
+
+  public async getValue(args: {[key: string]: any}) {
+    const account = args.account;
+    const region = args.region;
+    if (!this.isHostedZoneProps(args)) {
+      throw new Error(`HostedZoneProvider requires domainName property to be set in ${args.props}`);
+    }
+    const domainName = args.domainName;
+    debug(`Reading hosted zone ${account}:${region}:${domainName}`);
+    const r53 = await this.aws.route53(account, region, Mode.ForReading);
+    const response = await r53.listHostedZonesByName({ DNSName: domainName }).promise();
+    if (!response.HostedZones) {
+      throw new Error(`Hosted Zone not found in account ${account}, region ${region}: ${domainName}`);
+    }
+    const candidateZones = await this.filterZones(r53, response.HostedZones,
+      !!args.privateZone, args.vpcId);
+    if (candidateZones.length > 1) {
+      const filteProps = `dns:${domainName}, privateZone:${args.privateZone}, vpcId:${args.vpcId}`;
+      throw new Error(`Found more than one matching HostedZone ${candidateZones} for ${filteProps}`);
+    }
+    return candidateZones[0];
+  }
+
+  private async filterZones(
+    r53: AWS.Route53, zones: AWS.Route53.HostedZone[],
+    privateZone: boolean,
+    vpcId: string | undefined): Promise<AWS.Route53.HostedZone[]> {
+
+      let candidates: AWS.Route53.HostedZone[] = [];
+      if (privateZone) {
+        candidates = zones.filter(zone => zone.Config && zone.Config.PrivateZone);
+      } else {
+        candidates = zones.filter(zone => !zone.Config || !zone.Config.PrivateZone);
+      }
+      if (vpcId) {
+        const vpcZones: AWS.Route53.HostedZone[] = [];
+        for (const zone of candidates) {
+          await r53.getHostedZone({Id: zone.Id}, (err, data) => {
```

Comment:
I should have been more specific here. I'm not sure whether this works or not, but it sure looks weird to me. I actually meant doing the following:

```ts
const data = await r53.getHostedZone({ Id: zone. Id }).promise();
// No need to check for err, just look at data afterwards
```

It's the same pattern as for the `listHostedZonesByName` call you're doing in the other function, and basically the way all SDK calls should look.

The `.promise()` at the end is a weird convention that the SDK has in there (they could have made it so it works without it, but they didn't). It makes it possible to `await` this API call.